libraries/botframework-connector/src/auth/microsoftAppCredentials.ts

```diff
@@ -1,83 +1,156 @@
+        if (this.shouldSetToken(webResource)) {
+            let token = await this.getToken();
+            return new msrest.TokenCredentials(token).signRequest(webResource);
+        }
 
-            } else {
-                resolve(webResource);
+        return webResource;
+    }
+
+    async getToken(forceRefresh: boolean = false): Promise<string> {
+        if (!forceRefresh) {
+            // check the global cache for the token. If we have it, and it's valid, we're done.
+            let oAuthToken = MicrosoftAppCredentials.cache.get(this.tokenCacheKey);
+            if (oAuthToken) {
+                // we have the token. Is it valid?
+                if (oAuthToken.expiration_time > new Date(Date.now())) {
+                    return oAuthToken.access_token;
+                }
             }
-        });
+        }
+
+        // We need to refresh the token, because:
+        // 1. The user requested it via the forceRefresh parameter
+        // 2. We have it, but it's expired
+        // 3. We don't have it in the cache.
+        let oAuthToken = await this.refreshToken();
+        MicrosoftAppCredentials.cache.set(this.tokenCacheKey, oAuthToken);
+        return oAuthToken.access_token;
     }
 
-    private getAccessToken(cb: (err: Error, accessToken: string) => void): void {
-        if (!this.accessToken || new Date().getTime() >= this.accessTokenExpires) {
+    private refreshToken(): Promise<OAuthResponse> {
+        return new Promise<OAuthResponse>((resolve, reject) => {
             // Refresh access token
             var opt: request.Options = {
                 method: 'POST',
-                url: MicrosoftAppCredentials.refreshEndpoint,
+                url: this.oAuthEndpoint,
                 form: {
                     grant_type: 'client_credentials',
                     client_id: this.appId,
                     client_secret: this.appPassword,
-                    scope: MicrosoftAppCredentials.refreshScope
+                    scope: this.oAuthScope
                 }
             };
-            request(opt, (err: any, response :any, body: any) => {
+
+            request(opt, (err: any, response: any, body: any) => {
                 if (!err) {
                     if (body && response.statusCode && response.statusCode < 300) {
                         // Subtract 5 minutes from expires_in so they'll we'll get a
                         // new token before it expires.
-                        var oauthResponse = JSON.parse(body);
-                        this.accessToken = oauthResponse.access_token;
-                        this.accessTokenExpires = new Date().getTime() + ((oauthResponse.expires_in - 300) * 1000);
-                        cb(null!, this.accessToken);
+                        var oauthResponse = <OAuthResponse>JSON.parse(body);
+                        oauthResponse.expiration_time = new Date(Date.now() + (oauthResponse.expires_in * 1000) - 60000);
```

Comment:
>60000 [](start = 114, length = 5)

This is 1 minute not 5. I double checked the v3 code and we definitely subtract 5 minutes from this.

Reply:
fixed